Reconnect broken LDAP connections

zmb3 opened the pull request:
Ensure that our LDAP client returns trace.ConnectionProblem errors
for LDAP error 200 (network error).

When this condition is detected, we re-initialize our LDAP client
to get a fresh connection.

Also clean up some of the errors in the reconciler so that they
include the resources kind and name without stringify-ing the
entire resource.

Lastly, add both a connect and request timeout for our LDAP client,
allowing us to detect connection issues sooner.

Fixes #9808

zmb3 committed:
Reconnect broken LDAP connections

Ensure that our LDAP client returns trace.ConnectionProblem errors
for LDAP error 200 (network error).

When this condition is detected, we re-initialize our LDAP client
to get a fresh connection.

Also clean up some of the errors in the reconciler so that they
include the resources kind and name without stringify-ing the
entire resource.

Lastly, add both a connect and request timeout for our LDAP client,
allowing us to detect connection issues sooner.

Fixes #9808

zmb3 commented:
Tested this by lowering the reconcile period from 5 minutes to 1 minute and rebooting the domain controller while teleport is running.

```
2022-02-06T17:52:34-07:00 DEBU [WINDOWS_D] Reconciling 1 current resources with 1 new resources. services/reconciler.go:102
2022-02-06T17:52:34-07:00 INFO [WINDOWS_D] windows_desktop WIN-I44F9TN11M3-zacmanual-example-com updated, updating. services/reconciler.go:169

<reboot here>

2022-02-06T17:53:34-07:00 DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-06T17:53:54-07:00 INFO [WINDOWS_D] LDAP connection error when searching for desktops, reinitializing client desktop/discovery.go:130
2022-02-06T17:53:54-07:00 INFO [CA]        Generating TLS certificate 
2022-02-06T17:53:54-07:00 INFO [WINDOWS_D] Teleport CA already present in NTAuthStore in LDAP desktop/windows_server.go:877
2022-02-06T17:53:54-07:00 INFO [WINDOWS_D] Updated CRL for Windows logins via LDAP desktop/windows_server.go:937
2022-02-06T17:53:54-07:00 DEBU [WINDOWS_D] Reconciling 1 current resources with 1 new resources. services/reconciler.go:102
2022-02-06T17:53:54-07:00 DEBU [WINDOWS_D] windows_desktop WIN-I44F9TN11M3-zacmanual-example-com is already registered. services/reconciler.go:182

2022-02-06T17:54:34-07:00 DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-06T17:54:34-07:00 DEBU [WINDOWS_D] discovered 1 Windows Desktops desktop/discovery.go:138
```

rosstimothy commented on lib/srv/desktop/discovery.go:
```diff
@@ -122,7 +122,15 @@ func (s *WindowsService) getDesktopsFromLDAP() types.ResourcesWithLabels {
+		// in this case, so that the reconciler doesn't delete the desktops it already
+		// knows about.
+		s.cfg.Log.Info("LDAP connection error when searching for desktops, reinitializing client")
+		s.initializeLDAP()
```
Looks like this error isn't being handled. Logging it might be the only actionable thing to do here though.

zmb3 replied:
Yeah, I was thinking I would avoid spamming with the same error message in case this takes several retries before it succeeds, but I'm not opposed to logging it.

rosstimothy commented on lib/srv/desktop/ldap.go:
```diff
@@ -113,6 +114,8 @@ func (c *ldapClient) create(dn string, class string, attrs map[string][]string)
 			case ldap.LDAPResultInsufficientAccessRights:
 				return trace.AccessDenied("insufficient permissions to create %q: %v", dn, err)
+			case ldap.ErrorNetwork:
+				return trace.ConnectionProblem(err, "LDAP network error")
```
Looks like all other `trace.ConnectionProblem` that were added contain the `dn`, should we add it here as well?

zmb3 replied:
Will do.

ibeckermayer reviewed (commented):
Tested it locally and it worked for me as well.

rcanderson23 reviewed (approved):
echo @ibeckermayer that this fixed the issue in my dev environment

rcanderson23 commented:
May have jumped the gun. I see the reconnect succeed but I think the resource is still expiring (gone in the UI) and subsequent updates fail:
```
2022-02-07T17:14:18Z DEBU [WINDOWS_D] Changing state connecting -> connected. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:208
2022-02-07T17:14:18Z DEBU [WINDOWS_D] Discovery request channel opened: teleport-discovery. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:517
2022-02-07T17:14:18Z DEBU [WINDOWS_D] handleDiscovery requests channel. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:535
2022-02-07T17:14:18Z INFO [WINDOWS_D] Teleport CA already present in NTAuthStore in LDAP desktop/windows_server.go:877
2022-02-07T17:14:18Z INFO [WINDOWS_D] Updated CRL for Windows logins via LDAP desktop/windows_server.go:937
2022-02-07T17:14:33Z DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-07T17:14:33Z DEBU [WINDOWS_D] discovered 1 Windows Desktops desktop/discovery.go:138
2022-02-07T17:14:33Z DEBU [WINDOWS_D] resolved WIN-ONUR9EC95N5.ad.carsonanderson.net => [10.192.2.124] desktop/discovery.go:194
2022-02-07T17:14:33Z DEBU [WINDOWS_D] Reconciling 0 current resources with 1 new resources. services/reconciler.go:102
2022-02-07T17:14:33Z INFO [WINDOWS_D] windows_desktop WIN-ONUR9EC95N5-ad-carsonanderson-net matches, creating. services/reconciler.go:148
2022-02-07T17:19:18Z DEBU [WINDOWS_D] Ping -> 10.192.101.5:3024. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:479
2022-02-07T17:19:34Z DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-07T17:19:34Z DEBU [WINDOWS_D] discovered 1 Windows Desktops desktop/discovery.go:138
2022-02-07T17:19:34Z DEBU [WINDOWS_D] resolved WIN-ONUR9EC95N5.ad.carsonanderson.net => [10.192.2.124] desktop/discovery.go:194
2022-02-07T17:19:34Z DEBU [WINDOWS_D] Reconciling 1 current resources with 1 new resources. services/reconciler.go:102
2022-02-07T17:19:34Z INFO [WINDOWS_D] windows_desktop WIN-ONUR9EC95N5-ad-carsonanderson-net updated, updating. services/reconciler.go:169
2022/02/07 17:20:20 tls: failed to send closeNotify alert (but connection was closed anyway): write tcp 10.244.19.181:37592->10.192.2.124:636: write: broken pipe
2022-02-07T17:24:18Z DEBU [WINDOWS_D] Ping -> 10.192.101.5:3024. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:479
2022-02-07T17:24:22Z DEBU [WINDOWS_D] Attempting to restart watch after waiting 4.429096757s. resource-kind:lock services/watcher.go:211
2022-02-07T17:24:22Z DEBU [WINDOWS_D] Triggering scheduled refetch. resource-kind:lock services/watcher.go:220
2022-02-07T17:24:22Z DEBU [WINDOWS_D] Starting watch. resource-kind:lock services/watcher.go:182
2022-02-07T17:24:34Z DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-07T17:24:34Z INFO [WINDOWS_D] LDAP connection error when searching for desktops, reinitializing client desktop/discovery.go:130
2022-02-07T17:24:34Z INFO [WINDOWS_D] Teleport CA already present in NTAuthStore in LDAP desktop/windows_server.go:877
2022-02-07T17:24:34Z INFO [WINDOWS_D] Updated CRL for Windows logins via LDAP desktop/windows_server.go:937
2022-02-07T17:24:34Z DEBU [WINDOWS_D] Reconciling 1 current resources with 1 new resources. services/reconciler.go:102
2022-02-07T17:24:34Z DEBU [WINDOWS_D] windows_desktop WIN-ONUR9EC95N5-ad-carsonanderson-net is already registered. services/reconciler.go:182
2022-02-07T17:29:18Z DEBU [WINDOWS_D] Ping -> 10.192.101.5:3024. leaseID:1 target:dev-teleport.carsonanderson.net:3024 reversetunnel/agent.go:479
2022-02-07T17:29:34Z DEBU [WINDOWS_D] searching for desktops with LDAP filter (&(objectClass=computer)(!(objectClass=msDS-GroupManagedServiceAccount))) desktop/discovery.go:122
2022-02-07T17:29:34Z DEBU [WINDOWS_D] discovered 1 Windows Desktops desktop/discovery.go:138
2022-02-07T17:29:34Z DEBU [WINDOWS_D] resolved WIN-ONUR9EC95N5.ad.carsonanderson.net => [10.192.2.124] desktop/discovery.go:194
2022-02-07T17:29:34Z DEBU [WINDOWS_D] Reconciling 1 current resources with 1 new resources. services/reconciler.go:102
2022-02-07T17:29:34Z INFO [WINDOWS_D] windows_desktop WIN-ONUR9EC95N5-ad-carsonanderson-net updated, updating. services/reconciler.go:169
2022-02-07T17:29:34Z ERRO [WINDOWS_D] "desktop reconciliation failed: failed to update windows_desktop WIN-ONUR9EC95N5-ad-carsonanderson-net\n\twindows desktop \"WIN-ONUR9EC95N5-ad-carsonanderson-net\" doesn't exist" desktop/discovery.go:96
```

If this is a known issue than you can ignore the update as the ldap reconnection part seems to be working for me.

zmb3 commented on lib/srv/desktop/discovery.go:
```diff
@@ -122,7 +122,15 @@ func (s *WindowsService) getDesktopsFromLDAP() types.ResourcesWithLabels {
+		// in this case, so that the reconciler doesn't delete the desktops it already
+		// knows about.
+		s.cfg.Log.Info("LDAP connection error when searching for desktops, reinitializing client")
+		s.initializeLDAP()
```
Yes, will add some logging as Tim pointed out the same.

We're not "leaking" that goroutine, as it will be terminated when the service shuts down. We do still need to periodically renew our certificate, but I think what you're pointing out is that we can technically cancel the renewal we've already scheduled as this will schedule a new one.

I don't think there's an issue here though, as:

- there's nothing bad about renewing our certificate more frequently than necessary
- we only schedule a renewal if we successfully got a cert, so in cases where we're in an error loop we won't be scheduling a bunch of renewals.

Thoughts?

zmb3 replied:
@ibeckermayer check out the latest commit - I think it's a better way to address the renewal and should address your comment.

ibeckermayer replied:
> We're not "leaking" that goroutine, as it will be terminated when the service shuts down.

True, bad description of the problem on my part.

> check out the latest commit - I think it's a better way to address the renewal and should address your comment.

Yeah looks good.

zmb3 commented:
Thanks @rcanderson23 - I thought I had resolved all such instances of that, but the best option is probably to use upsert everywhere. Let me push a small revision and I'd love for you to give it another try.

zmb3 committed:
Reschedule pending renewals after succesful LDAP cert renewals

Prior to this change, if we encountered an LDAP network error and
re-initialized the LDAP client, we would schedule a new renewal
but also leave the existing renewal in place. Since each succesful
renewal schedules a subsequent renewal, we would be renewing twice
as often as necessary after recovering from this network error.
Each subsequent network error that we recover from would create
yet another set of periodic renewals. In the worst case we could
ultimately exhaust system resources by doing nothing but handling
LDAP cert renewals.

Fix this by using a single *time.Timer that is used to handle the
next LDAP cert renewal. When it comes time to schedule a renewal,
we use timer.Reset() to bump the existing renewal out rather than
schedule an additional one.

This introduced the need for a mutex, so I got rid of the atomic
load/store for ldapInitialized and put that behind the same mutex.

rosstimothy commented on lib/srv/desktop/windows_server.go:
```diff
@@ -519,6 +540,10 @@ func (s *WindowsService) startStaticHostHeartbeats() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.ldapCertRenew.Stop()
```
Is there any chance `s.ldapCertRenew` will be `nil` here? Can I `Close` before the first successful call to `initializeLDAP`?

zmb3 replied:
Good catch, it is possible. Unlikely, but possible. Fixed.

ibeckermayer commented on lib/srv/desktop/discovery.go:
```diff
@@ -113,7 +112,7 @@ func (s *WindowsService) ldapSearchFilter() string {
 // getDesktopsFromLDAP discovers Windows hosts via LDAP
 func (s *WindowsService) getDesktopsFromLDAP() types.ResourcesWithLabels {
-	if atomic.LoadInt32(&s.ldapInitialized) == 0 {
+	if !s.ldapReady() {
```
I think this check here has the potential to cause us issues:

Consider the case that in one reconcile loop,
`readWithFilter` (line 123 below this) fails and we call `s.initializeLDAP` again (line 130). If that `initializeLDAP` fails, then inside it we'll be setting `s.ldapInitialized = false`. So on the next reconcile loop, this check will be triggered and we'll return `nil`, but won't try to `initializeLDAP` again, and will do so for every subsequent reconcile loop.

zmb3 committed:
Add retries to LDAP client initialization

Prior to this change, we had retry logic on the initial setup and
on receiving a network error during a query. We did not have retry
logic in place if the re-initialization failed. This would put the
system in a state where desktop discovery stops running until the
cert approaches expiration (up to a maximum of 5-6 hours)

Now, we schedule the next re-initialization to be either:
- 10 minutes later, in the case of an error
- After 2/3 of the cert's validity has elapsed (5-6 hours),
  for the happy path
